### PR TITLE
TheoryIds for UF theory.

### PR DESCRIPTION
**Pull request by nafur**
This PR introduces new InferenceId for the uf theory and uses them instead of InferenceId::UNKNOWN.

### CODE REVIEW

File: src/theory/inference_id.h [hunk 1]
```diff
@@ -400,7 +400,23 @@ enum class InferenceId
...
+  UF_CARD_EQUIV,
+  UF_CARD_MONOTONE_COMBINED,
+  UF_CARD_SIMPLE_CONFLICT,
+  UF_CARD_SPLIT,
+  // HO collect model info application
+  UF_HO_APPLY_EXPAND,
+  UF_HO_APPLY_CONV_SKOLEM,
+  UF_HO_CMI_APP,
+  UF_HO_CMI_EXTENSIONALITY,
+  UF_HO_EXTENSIONALITY,
```
Comment (HanielB) [inline src/theory/inference_id.h:417]:
```suggestion
  //-------------------------------------- begin HO extension to UF
  // Encodes an n-ary application as a chain of binary HO_APPLY applications
  //   (= (f t1 ... tn) (@ (@ ... (@ f t1) ...) tn))
  UF_HO_APP_ENCODE,
  UF_HO_APP_CONV_SKOLEM,
  // Adds an extensionality lemma to witness that disequal functions have
  // different applications
  //   (not (= (f sk1 .. skn) (g sk1 .. skn))
  UF_HO_EXTENSIONALITY,
  //-------------------------------------- begin model-construction specific part
  // These rules are necessary to ensure that we build models properly. For more
  // details see Section 3.3 of Barbosa et al. CADE'19.
  //
  // Enforces that a regular APPLY_UF term in the model is equal to its HO_APPLY
  // equivalent by adding the equality as a lemma
  //   (= (f t1 ... tn) (@ (@ ... (@ f t1) ...) tn))
  UF_HO_MODEL_APP_ENCODE,
  // Adds an extensionality lemma to witness that disequal functions have
  // different applications
  //   (not (= (f sk1 .. skn) (g sk1 .. skn))
  UF_HO_MODEL_EXTENSIONALITY,
  //-------------------------------------- end model-construction specific part
  //-------------------------------------- end HO extension to UF
```

**Review by HanielB — COMMENTED** [review]
I defer to @ajreynol for the cardinality part and for UF_HO_APP_CONV_SKOLEM. I added some documentation and minor changes to the names.

File: src/theory/inference_id.h (hunk 1 above)
Comment (nafur) [reply src/theory/inference_id.h]: Thanks!

**Review by ajreynol — APPROVED** [review]
I can make a pass to add documentation separately.

**Pull request merged**